### ISSUE TITLE
TST/MAINT: ndimage: follow-ups for array API standard support

### ISSUE THREAD

**Issue by lucascolley** [scipy/scipy#21280]
Follow-ups from gh-21150:

- [ ] https://github.com/scipy/scipy/pull/21091#discussion_r1676935699: We should try to add an `@xfail_xp_backends` decorator, to match `@skip_xp_backends`.
- [ ] https://github.com/scipy/scipy/pull/21150#issuecomment-2239154842: Tests for array-likes should be added throughout.
- [ ] https://github.com/scipy/scipy/pull/21150#discussion_r1693961428: We should try to add the ability to say "only run for these backends" to the skip infra.
- [ ] https://github.com/scipy/scipy/pull/21150#discussion_r1694278963: Some functions produce object arrays which are incompatible with alternative backends. We probably want to deprecate and return regular arrays instead.

**Comment by shivanimall** [comment]
Thanks @lucascolley, I am hoping to pass in PyTorch tensors (on GPU) to `scipy.ndimage.map_coordinates`, currently this fails

**Comment by shivanimall** [comment]
Currently, PyTorch has an existing implementation of [grid sample](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html) which is similar to `map_coordinates` however, it changes the range of coordinates to -1 to 1. I would like to avoid this normalisation, if possible. So, I resorted to using scipy's functionality for the same. 

Would this mean that I would have to build the code with torch.compile since there is no direct support for `map_coordinates` in torch?

**Comment by lucascolley** [comment]
If you want to keep your arrays on the GPU, then I don't think `torch.compile` can help here, given that `ndimage.map_coordinates` is implemented on NumPy arrays which are CPU-only.

**Comment by shivanimall** [comment]
thanks @lucascolley, would modifying any relevant source code for `scipy.ndimage.map_coordinates` to accept PyTorch tensors be an option?

**Comment by jakirkham** [comment]
Are there still things to do here? Looks like all the items listed above are merged PRs

**Comment by lucascolley** [comment]
I think some of the unresolved items are linking to (unaddressed) comments from merged PRs

**Comment by ev-br** [comment]
> - [ ] Some functions produce object arrays which are incompatible with alternative backends. We probably want to deprecate and return regular arrays instead.

This item, turns out, is not related to array types. It's just an API quirk, where calling [ndimage.histogram](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.histogram.html) with `index` being a sequence returns an object array:

```
In [10]: labels = np.asarray([1, 0, 1, 1, 2, 2, 2, 2])

In [11]: input = np.asarray([1, 1, 3, 4, 3, 5, 3, 3])

In [12]: ndimage.histogram(input, 0, 4, 5, labels, (1, 2))
Out[12]: array([array([0, 1, 0, 1, 1]), array([0, 0, 0, 3, 0])], dtype=object)

In [13]: ndimage.histogram(input, 0, 4, 5, labels, 1)
Out[13]: array([0, 1, 0, 1, 1])

In [14]: ndimage.histogram(input, 0, 4, 5, labels, 2)
Out[14]: array([0, 0, 0, 3, 0])

In [15]: scipy.__version__
Out[15]: '1.16.3'
```

The affected test is https://github.com/scipy/scipy/blob/v1.18.0/scipy/ndimage/tests/test_measurements.py#L1421-L1432 


The relevant source line traces back to at least 2010: https://github.com/scipy/scipy/commit/48cbdfea4e78cc022af5676788af282a8ce9e46e#diff-4753ca1b1d654cd749faee3a2e809f90b4f882a45b42bff1175688a75f2be9adR603

Conclusion: is unrelated to anything Array API; if we want to change the return type, this should go through a usual process for API changes. First and foremost, it needs a champion to argue for the API change (however ugly the API is). 


EDIT: A separate issue for a separate issue: https://github.com/scipy/scipy/issues/25555

**Comment by ev-br** [comment]
> Tests for array-likes should be added throughout.

Now that's easier with AI assistants: https://github.com/scipy/scipy/pull/25554

**Comment by rgommers** [comment]
All done now that gh-25554 is merged, so closing. Thanks all!
